state/machine.go

```diff
@@ -4,325 +4,524 @@ import (
+		if propose.SignedPropose.Block.Height > machine.currentHeight {
+			if propose.LastCommit == nil {
+				return nil
+			}
+			machine.SyncCommit(*propose.LastCommit)
+			machine.currentRound = propose.Round()
+		}
+	}
+	switch machine.currentState.(type) {
 	case WaitingForPropose:
 		return machine.waitForPropose(transition)
 	case WaitingForPolka:
 		return machine.waitForPolka(transition)
 	case WaitingForCommit:
 		return machine.waitForCommit(transition)
 	default:
-		panic(fmt.Errorf("unexpected state type %T", machine.state))
+		panic(fmt.Errorf("unexpected state type %T", machine.currentState))
 	}
 }
 
 func (machine *machine) waitForPropose(transition Transition) Action {
 	switch transition := transition.(type) {
 	case Proposed:
-		// FIXME: Proposals can (optionally) include a Polka to encourage
-		// unlocking faster than would otherwise be possible.
-		machine.state = WaitingForPolka{}
-		return machine.preVote(&transition.Block)
+		if machine.shard.Leader(machine.currentRound).Equal(transition.Signatory) {
+			if transition.ValidRound < 0 {
+				machine.ticksAtProposeState = -1
+				machine.currentState = WaitingForPolka{}
+				if machine.lockedRound == -1 || machine.lockedValue.Block.Equal(transition.Block.Block) {
+					return machine.broadcastPreVote(&transition.Block)
+				}
+				return machine.broadcastPreVote(nil)
+			}
+			if polka, polkaRound := machine.polkaBuilder.Polka(machine.currentHeight, machine.shard.ConsensusThreshold()); polkaRound != nil {
+				if polka.Block != nil && polka.Block.Block.Equal(transition.Block.Block) && transition.ValidRound < machine.currentRound {
+					machine.ticksAtProposeState = -1
+					machine.currentState = WaitingForPolka{}
+					if machine.lockedRound <= transition.ValidRound || machine.lockedValue.Block.Equal(transition.Block.Block) {
+						return machine.broadcastPreVote(&transition.Block)
+					}
+					return machine.broadcastPreVote(nil)
+				}
+			}
+		}
 
 	case PreVoted:
-		_ = machine.polkaBuilder.Insert(transition.SignedPreVote)
+		machine.polkaBuilder.Insert(transition.SignedPreVote)
 
 	case PreCommitted:
-		_ = machine.commitBuilder.Insert(transition.SignedPreCommit)
-
-	case TimedOut:
-		machine.state = WaitingForPolka{}
-		return machine.preVote(nil)
+		machine.commitBuilder.Insert(transition.SignedPreCommit)
+
+	case Ticked:
+		machine.ticksAtProposeState++
+		maxTicksToTimeout := int(NumTicksToTriggerTimeOut + machine.currentRound)
+		if machine.ticksAtProposeState > maxTicksToTimeout {
+			machine.ticksAtProposeState = -1
+			machine.currentState = WaitingForPolka{}
+			return machine.broadcastPreVote(nil)
+		}
 
 	default:
 		panic(fmt.Errorf("unexpected transition type %T", transition))
 	}
 
-	return machine.checkCommonExitConditions()
+	return nil
 }
 
 func (machine *machine) waitForPolka(transition Transition) Action {
+	machine.checkAndSchedulePreVoteTimeout()
+
+	polka := &block.Polka{}
+
 	switch transition := transition.(type) {
```

Comment:
Got up to here for today, will pick up tomorrow.

Question (reminder for myself to ask): I don't see where line 47 in the paper is accounted for in the propose and prevote steps.